### PR TITLE
File handle management

### PR DESCRIPTION
**Pull request by HammadB**
We keep file handles open for the lifetime of an index, this can exhaust system file descriptor limits in the case of large numbers of indexes on systems with low file descriptor limits. This PR adds a way to open/close the needed file descriptors for calling python code to manage overall file handle usage across indexes.

### CODE REVIEW

File: python_bindings/bindings.cpp [hunk 1]
```diff
@@ -146,6 +146,7 @@ template<typename dist_t, typename data_t = float>
 class Index {
  public:
     static const int ser_version = 1;  // serialization version
+    static const int file_handle_count = 4; // number of file handles to open for persistent index
```
Comment (atroyn) [inline python_bindings/bindings.cpp:149]: Is this just so we can read it from the python bindings? How would we use this / how would a user use this?

Comment (atroyn) [reply python_bindings/bindings.cpp:149]: Makes me think we ought to refactor `setupPersistentIndexFileHandles` to be a const array of const structs with members and locations we can pass to a single method, but that's probably overkill for this PR.

**Review by atroyn — APPROVED** [review]

File: python_bindings/bindings.cpp (hunk 1 above)
Comment (HammadB) [reply python_bindings/bindings.cpp:149]: Yeah its just so the python code is not hardcoding the file count.

I agree with that, I also think the multi-index management ought to be shoved into c++.

Comment (atroyn) [reply python_bindings/bindings.cpp:149]: Why does python need to know about the file count?

Comment (HammadB) [reply python_bindings/bindings.cpp:149]: In https://github.com/chroma-core/chroma/pull/891 we allow python to manage multiple index file handle status in order to buffer against OS limits.

**Pull request merged**